onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -30,13 +32,35 @@
 namespace onnxruntime {
 namespace session_state_utils {
 
+static common::Status ReserveUsingArenaFromShapeAndType(const TensorShape& tensor_shape, const DataTypeImpl* type,
+                                                        const AllocatorPtr& arena_alloc, /*out*/ void*& p_data) {
+  if (arena_alloc->Info().alloc_type != OrtArenaAllocator)
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Given allocator must be an arena based allocator");
+
```

Comment:
This probably doesn't matter, but if it's not an arena based allocator we could just call Alloc couldn't we?

Reply:
Actually that seems to currently be handled elsewhere, but that means we're checking if it's arena based in multiple places instead of maybe just doing the check once here and falling back to using Alloc is necessary.

Reply:
As far as I can see, we only do the check once - Calling Alloc() is done within the existing Tensor ctor already. We are special casing the "reserve" option and passing the buffer allocated to the new Tensor ctor with a deleter.

Reply:
Fixed based on offline sync

Reply:
Fixed based on offline sync